cpp/dolfinx/fem/assemble_vector_impl.h

```diff
@@ -26,6 +26,43 @@
 namespace dolfinx::fem::impl
 {
 
+/// Extract a sub-set of entities whose cell index (local to process) is marked
+///
+/// @param[in] entities The entities.
+/// @param[in] cell_indicator An indicator vector. The size of cell_indicator
+/// should be smaller than max(entities).
+/// @tparam stride  The number of entries in each entity (1 for cells, 2 for
+/// exterior facets, 4 for interior facets)
+/// @returns A list of sub-entities
+template <std::size_t stride>
+std::vector<std::int32_t>
+_extract_entities(std::span<const std::int32_t> entities,
+                  std::span<const std::int8_t> cell_indicator)
+{
+  std::vector<std::int32_t> sub_entities;
+  sub_entities.reserve(entities.size());
+  for (std::size_t e = 0; e < entities.size(); e += stride)
+  {
+    if constexpr (stride == 1)
```

Comment:
And it would never happen, as we only have three types of entities, cells (stride 1), exterior facets (stride 2), interior facets (stride 4), as mentioned in the `@tparam` documentation.

Reply:
The documentation doesn't impose any constraints. Constraints should be checked in the code when possible. 

In any case, `constexpr` seems like an unnecessary complication. It would be simpler to handle each case separately.